Support for expression converting types

lorenzo opened the pull request:
This adds the ability for Type classes to implement `toExpression()`, which was a missing feature for the Typing system to easily support complex types such as a Geo Point.

Closes #8041

lorenzo committed:
Starting to convert values to expressions when the type class offers this

lorenzo committed:
Making Comparison correctly traverse expressions in array values
Started testing the type conversion to expressions

lorenzo committed:
Implmented ValueBinder::generateMany() as a way to reduce funcitons calls
Also calling bindValue() directly instead of the bind() proxy for the same
reaason.

lorenzo committed:
Added tests for all expression that are capable of accepting
values convertible to other expressions

lorenzo committed:
Adding doc blocks and fixing some failing tests

markstory commented on src/Database/Expression/Comparison.php:
```diff
@@ -229,13 +263,54 @@ protected function _bindValue($value, $generator, $type)
+            unset($value[$k]);
+        }
+
+        $parts = $expressions;
```
This looks redundant.

markstory commented on src/Database/Expression/Comparison.php:
```diff
@@ -229,13 +263,54 @@ protected function _bindValue($value, $generator, $type)
     }
+
+    /**
+     * Returns an array with the original $values in the first poisition
```
Typo here.

markstory commented on src/Database/Type/TypeExpressionCasterTrait.php:
```diff
@@ -0,0 +1,62 @@
+ * if the type they should be converted to implements ExpressionTypeInterface
+ *
+ */
+trait TypeExpressionCasterTrait
```
Would `ExpressionTypeCaster` be a fitting name?

lorenzo replied:
`ExpressionTypeCasterTrait ` it is

markstory commented on src/Database/Type/TypeExpressionCasterTrait.php:
```diff
@@ -0,0 +1,62 @@
+            $result = [];
+            foreach ($value as $k => $v) {
+                $result[$k] = $converter->toExpression($v);
+            }
```
Could `array_map` be used here?

lorenzo replied:
Maybe, if it is not significantly slower I can change that

lorenzo replied:
array_map is actaully faster :)

markstory replied:
Yay.

markstory commented on tests/TestCase/Database/ExpressionTypeCastingTest.php:
```diff
@@ -0,0 +1,171 @@
+ * using the type classes
+ *
+ */
+class ExpressionTypeCastingTest extends TestCase
```
Should we also have some integration tests at the database package level?

lorenzo replied:
Yes, I'll add some

markstory replied:
Another case that would be good to cover is a use case of a Type class that maps a column to a Value object and then back into an expression. I know that was one of the big asks in #8041

markstory committed:
Fix typos in doc blocks.

lorenzo committed:
Renamed trait

lorenzo committed:
Using the new trait and simplifying code

jippi commented on src/Database/Expression/Comparison.php:
```diff
@@ -52,6 +54,21 @@ class Comparison implements ExpressionInterface, FieldInterface
+     *
+     * @var bool
+     */
+    protected $_isMultiple = false;
```
`_isTraversable` ?

lorenzo committed:
Making sure ValuesExpression also casts values to expressions

lorenzo committed:
It turns out that SQLite 3.7.11+ implements multi-inserts!

Version 3.8 is the minimum sqlite version fro php 5.5

lorenzo committed:
Adding integration test for the Type to expression casting

lorenzo committed:
Simplifying code

lorenzo commented:
@markstory Added some integration tests

lorenzo commented on src/Database/Dialect/SqliteDialectTrait.php:
```diff
@@ -150,59 +150,6 @@ protected function _transformFunctionExpression(FunctionExpression $expression)
-     * @param \Cake\Database\Query $query The query to translate
-     * @return \Cake\Database\Query
-     */
-    protected function _insertQueryTranslator($query)
```
Thanks PHP 5.5 for linking to SQLite 3.8!

markstory replied:
So good. This was an interesting work around :smile:

lorenzo committed:
Making tests pass in postgresql

lorenzo committed:
Adding a function traslator for the SUBSTR function in SQL Server

Improving the function translator for SUBSTR in SQL Server

Appending a LEN function instead of using RIGHT

Trying to fix funciton translaotr

trying to fix sql server tests

one last try

Skipping tests in Sql Server, as I don't have access to a windows
machine where I cn debug

Doing the previous commit right

lorenzo commented:
I could not make the tests pass in SQL Server, since the tests I wrote use functions that are present in all other drivers except this one. Based on the error messages in appveyor I can see that the values are being casted to expressions correctly, I could just not find the way of making the test completely agnostic.

If anyone with a windows machine can give me a hand, I would appreciate it :)

lorenzo committed:
Fixed CS

markstory commented on src/Database/Type/ExpressionTypeCasterTrait.php:
```diff
@@ -0,0 +1,79 @@
+            $object = Type::build($type);
+            if ($object instanceof ExpressionTypeInterface) {
+                $result[$k] = $object;
+            }
```
Could this check be done as part of the `array_filter`?

```php
$types = array_filter($types, function ($type) {
  return (Type::build($type) instanceof ExpressionTypeInterface);
});
return array_map('Type::build', $types);
```

I've not run this code, but it might work.

lorenzo replied:
I think it could work

lorenzo replied:
On second look, I think I prefer this. In the filter I would need an `if` to remove empty strings and then check for the instance. So the code grows to basically the same size.

markstory replied:
Boourns.

markstory commented on tests/Fixture/OrderedUuidItemsFixture.php:
```diff
@@ -0,0 +1,45 @@
+use Cake\TestSuite\Fixture\TestFixture;
+
+/**
+ * Class BinaryUuiditemFixture
```
Doc tag is out of sync with class name.

markstory commented on tests/Fixture/OrderedUuidItemsFixture.php:
```diff
@@ -0,0 +1,45 @@
+     * @var array
+     */
+    public $fields = [
+        'id' => ['type' => 'string', 'length' => 32],
```
Could we use the UUID abstract type here?

lorenzo replied:
No, I specifically want to show that the ordered_uuid is doing something extra on top of the base string.

lorenzo replied:
Also uuid is 36 chars, I have chose 32 in order to show that dashes are removed

markstory replied:
Ok.

markstory commented on tests/TestCase/Database/ExpressionTypeCastingIntegrationTest.php:
```diff
@@ -0,0 +1,209 @@
+        return new FunctionExpression(
+            'CONCAT',
+            [$substr(15, 4), $substr(10, 4), $substr(1, 8), $substr(20, 4), $substr(25)]
+        );
```
Fancy!

markstory committed:
Fix up typos and add doc blocks.

markstory commented on tests/TestCase/Database/ExpressionTypeCastingIntegrationTest.php:
```diff
@@ -0,0 +1,209 @@
+        $result = $this->connection->newQuery()
+            ->select('id')
+            ->from('ordered_uuid_items')
+            ->where(['id' => '48298a29-81c0-4c26-a7fb-413140cf8569'], ['id' => 'ordered_uuid'])
```
Can conditions use Value objects as well?

lorenzo replied:
Yes, but that will require special code in the Type class... not sure that is interesting to see, as it has been shown many times for example with the DateTimeType

markstory replied:
I guess the scenario I was interested in covering was:

1. An entity is fetched from the DB.
2. Its properties are hydrated into value objects,
3. Value object property is used for useful things
  1. Querying other records.
  2. Being updated and re-persisted.

I think 1, 2, 3.2 are covered by the test cases so far.

markstory commented on tests/TestCase/Database/ExpressionTypeCastingIntegrationTest.php:
```diff
@@ -0,0 +1,209 @@
+                return $exp->between(
+                    'id',
+                    '482b7756-8da0-419a-b21f-27da40cf8569',
+                    '48298a29-81c0-4c26-a7fb-413140cf8569',
```
Can value objects be use here?

lorenzo replied:
Would not be interesting, imo. The thing I'm showing here is that passing a random uuid, can be converted to a set of expressions to remove the dashes and order it.

lorenzo replied:
I can add a test like that if you still think is relevant

markstory replied:
Ok if this test is for something else then :+1:

markstory commented:
I pushed up a commit with some docblock/typo fixes.

lorenzo commented:
@markstory Thanks for fixing my typos!

lorenzo committed:
More CS fixes

markstory commented:
Do we need any changes to handle value objects in any of the following places:

* GROUP BY
* ORDER BY
* WHERE clauses
* SELECT columns?

lorenzo commented:
@markstory SELECT, ORDER BY and GROUP BY require column names or expressions, not values. That's why I left them off initially, it is easy to wrap the value inside an expression object and let the automatic conversion work for you.

We do need them in WHERE clauses, and that's already tested. Again if you want the raw value put the, I'd say it is better to wrap in an expression, at least until the use cases for this feature come up a bit clearer.

markstory commented:
Ok. I was thinking of the case where someone wanted to select a constant value. If there is a way to do that with expression wrappers that's fine.

markstory commented:
@lorenzo Is there anything else that needs to be done on this beyond documentation updates?

lorenzo committed:
Added test for value object in conditions

lorenzo commented:
I will be adding an extra test case as you wanted, but did not have the time during the week. Want to help me document this?

markstory commented:
@lorenzo I'm happy to help with the docs.

lorenzo commented:
@markstory thanks!

lorenzo commented:
@markstory I added the additional test

markstory commented:
Looks good :+1:. I'll get started on the docs tonight.

lorenzo commented:
I'll merge, then :)

markstory commented:
Nice work @lorenzo :clap:
